lib/cloudregister/registerutils.py

```diff
@@ -1398,6 +1405,44 @@ def write_framework_identifier(cfg):
         framework_file.write(json.dumps(identifier))
 
 
+# ----------------------------------------------------------------------------
+def get_cached_instance_data():
+    """Get cached instance data if cache has not expired."""
+    try:
+        if instance_data_cache_expired():
+            return None
+        else:
+            cache_instance_data_path = os.path.join(
+                get_state_dir(), 'cache_instance_data'
+            )
+            with open(cache_instance_data_path) as f:
+                cached_instance_data = f.read()
+            return cached_instance_data
```

Comment:
This is not good enough, it opens the system to attack by replacing the `cache_instance_data` file. We'd have to prevent this by storing a checksum somewhere or doing some other tricks. Storing such state across invocations is tricky. It may be better to store the data in a tmpfile such that the cache is available within one invocation, i.e. the same process rather than trying to be available across invocations.

Reply:
Ack, I will change it to be available per invocation instead of several invocations

I understand the concern and I agree that is serious/worth changing the approach
That said, if someone wanted to fiddle with the metadata, they can do that now, right ?
I'm not saying we should not worry about this case, I'm saying that concern is valid for
other cases

Reply:
If someone wanted to inject suitable metadata into the process now they would have to intercept the traffic to the metadata server, i.e. they need something that intercepts `http://169.254.169/254` and then they'd need to implement the suitable API. That's a lot of work. Compared to

- Run 1000 instances A (BYOS)
- Run 1 instance B (PAYG)
- Every 4 minutes rsync the metadata from instance B to 1000 instances of A into the cache file location

The later approach I can implement in about 20 minutes.